### PR TITLE
Settings 2.0

### PR DESCRIPTION
**Pull request by MeguminSama**
Long awaited plugin settings page :3

### CODE REVIEW

**Comment by ArjixWasTaken** [comment]
Why do you delete `isStaff`?
image

**Comment by Vendicated** [comment]
merged into experiments

File: src/components/PluginSettings/PluginModal.tsx [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+    }
+
+    return (
+        <FormSection>
+            <FormTitle>{setting.name}</FormTitle>
+            <Select
+                isDisabled={setting.disabled?.() ?? false}
+                options={setting.options}
+                placeholder={setting.placeholder ?? "Select an option"}
+                maxVisibleItems={5}
```
Comment (Vendicated) [inline src/components/PluginSettings/PluginModal.tsx:190]: what happens if there's more than 5 options?

File: src/components/PluginSettings/PluginModal.tsx [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+                closeOnSelect={true}
+                select={onChange}
+                isSelected={v => v === state}
+                serialize={v => String(v)}
+                {...setting.componentProps}
+            />
+        </FormSection>
+    );
+}
+
```
Comment (Vendicated) [inline src/components/PluginSettings/PluginModal.tsx:282]: this file is kinda long, maybe split the component into individual files

File: src/components/PluginSettings/index.tsx [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+                    o[dep].push(plugin);
+                }
+            }
+        }
+        return o;
+    }, []);
+
+    const sortedPlugins = React.useMemo(() => Object.values(Plugins)
+        .sort((a, b) => a.name.localeCompare(b.name))
+        .sort((a, b) => a.required === b.required ? 0 : b.required ? -1 : 1), []);
```
Comment (Vendicated) [inline src/components/PluginSettings/index.tsx:120]: this double sort is kind of confusing. Does this keep the alphabetical order if required === required?

File: src/plugins/experiments.tsx [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+                match: /hasFreePremium=function\(\){return this.is Staff\(\)\s*\|\|/,
+                replace: "hasFreePremium=function(){return ",
+            },
+        ],
+    }],
+    settings: [
+        {
+            key: "enableIsStaff",
+            name: "Enable isStaff (requires restart)",
+            type: "boolean",
```
Comment (Vendicated) [inline src/plugins/experiments.tsx:49]: imo this should be an enum

File: src/plugins/experiments.tsx [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+                match: /return\s*(\w+)\.hasFlag\((.+?)\.STAFF\)}/,
+                replace: "return Vencord.Webpack.Common.UserStore.getCurrentUser().id===$1.id||$1.hasFlag($2.STAFF)}"
+            },
+            {
+                match: /hasFreePremium=function\(\){return this.is Staff\(\)\s*\|\|/,
+                replace: "hasFreePremium=function(){return ",
+            },
+        ],
+    }],
+    settings: [
```
Comment (Vendicated) [inline src/plugins/experiments.tsx:45]: why not 
```js
settings: {
   enableIsStaff: {
   }
}
```

File: src/utils/types.ts [hunk 6]
```diff
@@ -41,6 +42,33 @@ interface PluginDef {
...
+    aboutComponent?(): React.ReactNode;
+}
+
+export interface PluginSettingsItem {
+    /** Unique key of the setting. Try not to change this. */
+    key: string;
+    name: string;
+    type: "string" | "number" | "boolean" | "select";
+    /** On selects, use default key in .options instead of this */
+    default?: string | number | boolean;
```
Comment (Vendicated) [inline src/utils/types.ts:55]: could make these strongly typed based on .type with some ts fuckery

File: src/utils/types.ts [hunk 7]
```diff
@@ -41,6 +42,33 @@ interface PluginDef {
...
+    placeholder?: string;
+    options?: PluginSettingsItemOption[];
+    onChange?(newValue: any): void;
+    disabled?(): boolean;
+    restartNeeded?: boolean;
```
Comment (Vendicated) [inline src/utils/types.ts:60]: maybe a target: option (desktop, web, both) could be useful here?

File: src/utils/types.ts [hunk 8]
```diff
@@ -41,6 +42,33 @@ interface PluginDef {
...
+export interface PluginSettingsItemOption {
+    label: string;
+    value: string | number | boolean;
+    default?: boolean;
 }
 
 export type IpcRes<V = any> = { ok: true; value: V; } | { ok: false, error: any; };
+
+export enum PluginCategory {
+    Utility = "Utility",
```
Comment (Vendicated) [inline src/utils/types.ts:73]: what's this for? :o

File: src/components/PluginSettings/PluginModal.tsx (hunk 1 above)
Comment (MeguminSama) [reply src/components/PluginSettings/PluginModal.tsx:190]: I believe it's just the length of the dropdown component. Overflowed items will scroll.

File: src/components/PluginSettings/index.tsx (hunk 3 above)
Comment (MeguminSama) [reply src/components/PluginSettings/index.tsx]: it does keep the sorting if required === required. Returning `0` from sort will keep items in the same place.

File: src/utils/types.ts (hunk 8 above)
Comment (MeguminSama) [reply src/utils/types.ts:73]: Was planning categories. Will remove for now and work on categories after we merge this stuff first :3

**Comment by ArjixWasTaken** [comment]
Progress so far?
AKA when merge?

**Comment by MeguminSama** [comment]
@ArjixWasTaken soon. It's pretty much done. Just doing final touches.

File: src/plugins/experiments.tsx [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+            },
+            {
+                match: /hasFreePremium=function\(\){return this.is Staff\(\)\s*\|\|/,
+                replace: "hasFreePremium=function(){return ",
+            },
+        ],
+    }],
+    options: {
+        enableIsStaff: {
+            name: "Enable isStaff (requires restart)",
```
Comment (Vendicated) [inline src/plugins/experiments.tsx:47]: wouldn't it make more sense to call this description?

File: src/utils/types.ts [hunk 10]
```diff
@@ -34,13 +35,101 @@ interface PluginDef {
...
+    /**
+     * Whether this plugin is required and forcefully enabled
+     */
     required?: boolean;
     /**
      * Set this if your plugin only works on Browser or Desktop, not both
      */
     target?: "WEB" | "DESKTOP" | "BOTH";
+    /**
+     * Allows
```
Comment (Vendicated) [inline src/utils/types.ts:57]: Allows

Comment (MeguminSama) [reply src/utils/types.ts:57]: You love vscode refactor randomly deleting shit

File: src/webpack/common.tsx [hunk 11]
```diff
@@ -82,6 +86,10 @@ export const Toasts = {
     }
 };
 
+export const UserUtils = {
+    getUser: lazyWebpack(filters.byCode(".USER(", "getUser")),
```
Comment (Vendicated) [inline src/webpack/common.tsx:90]: maybe rename to fetchUser or userGet? I know discord themselves name this getUser but that name is more self explanatory. also type would be cool 
```suggestion
    getUser: lazyWebpack(filters.byCode(".USER(", "getUser")) as (id: string) => Promise<User>,
```

**Review by Vendicated — APPROVED** [review]

**Pull request merged**